README
** INTRODUCTION **

Dmidecode reports information about your system's hardware as described in
your system BIOS according to the SMBIOS/DMI standard. This information
typically includes system manufacturer, model name, serial number, BIOS
version, asset tag as well as a lot of other details of varying level of
interest and reliability depending on the manufacturer. This will often
include usage status for the CPU sockets, expansion slots (e.g. AGP, PCI,
ISA) and memory module slots, and the list of I/O ports (e.g. serial,
parallel, USB).

Part of the dmidecode code can be found in the Linux kernel, where DMI data
is used to enable or disable specific portions of code depending on the
specific hardware. Thus, one use of dmidecode is for kernel developers to
detect system "signatures" and add them to the kernel source code when
needed.

Beware that DMI data have proven to be too unreliable to be blindly trusted.
Dmidecode does not scan your hardware, it only reports what the BIOS told it
to.


** INSTALLATION **

The home web page for dmidecode is hosted on Savannah:
  http://www.nongnu.org/dmidecode/
You will find the latest version (including CVS) there, as well as fresh news
and other interesting material, such as a list of related projects and
articles.

This program was first written for Linux, and has since been reported to work
on FreeBSD, NetBSD, OpenBSD, BeOS, Cygwin and Solaris as well.

There's no configure script, so simply run "make" to build dmidecode, and
"make install" to install it. You also can use "make uninstall" to remove
all the files you installed. By default, files are installed in /usr/local
but you can change this behavior by editing the Makefile file and setting
prefix to wherever you want. You may change the C compiler and the
compilation flags as well.

Optionally, you can run "make strip" prior to "make install" if you want
smaller binaries. However, be aware that this will prevent any further
attempt to debug the programs.

Two parameters can be set in the Makefile file to make dmidecode work on
non-i386 systems. They should be used if your system uses the big endian
byte ordering (Motorola) or doesn't support unaligned memory accesses,
respectively. For example, compiling for a SPARC processor would require
both (but I am not aware of SPARC-based systems implementing SMBIOS).
Compiling for an IA64 processor requires the memory alignment workaround,
and it is enabled automatically.


** DOCUMENTATION **

Each tool has a manual page, found in the "man" subdirectory. Manual pages
are installed by "make install". See these manual pages for command line
interface details and tool specific information.

For an history of the changes made to dmidecode, see the CHANGELOG file.

If you need help, your best chances are to visit the web page (see the
INSTALLATION section above) or to get in touch with the developers directly.
Have a look at the AUTHORS file and contact one of the maintainers.

If you want to help with the development of dmidecode, please consider
joining the dmidecode-devel discussion list:
  http://lists.nongnu.org/mailman/listinfo/dmidecode-devel


** COMMON PROBLEMS **

MODEL SPECIFIC ISSUES

Dmidecode used not to work on IBM T-series laptops nor on Fujitsu-Siemens
S-series laptops under Linux. This was due to the fact that the DMI table
is at a memory location we couldn't seem to reach through /dev/mem. Although
I believe this reveals a problem in the Linux kernel, which should be fixed,
a workaround was found by Chad Smith, and we are using it for now. See the
IA-64 subsection below for more details. I posted about this problem on the
LKML but never received any answer. Since then, the list of affected systems
was extended to IBM X-series laptops and xSeries x445 servers.

IA-64

Dmidecode used to have problems on IA-64 systems. The first reason for
this is that dmidecode accesses the DMI table through /dev/mem, and reading
this file on an IA-64 system sometimes leads to a crash. A second reason is
that the method for locating the above-mentioned table differs on IA-64
(compared to x86), so dmidecode was likely to miss the table entry point.
This complex issue was reported by Glen Foster and Chad Smith from HP. We
have since been working on a solution, and dmidecode now supports IA-64
systems. Chad Smith noticed that, for some obscure reason, accessing the
/dev/mem file using mmap() instead of read() would work. Then, he wrote a
patch to export the DMI table address from the internal EFI table to /proc,
so dmidecode doesn't have to scan /dev/mem for it anymore. This patch was
since integrated into the main ia64 patch. Finally, I added the required
code to make it all work. So, in order to have dmidecode work on your IA-64
system, you need two things:
 - dmidecode version 2.2 or later;
 - an ia64 patched 2.4 kernel, using linux-2.4.21-ia64-030702.diff or any
   later version, or a 2.6 kernel.
Chad Smith tested dmidecode successfully on two different IA-64 systems, and
success has been reported by other users too. Non-Linux systems are not
supported.

MMAP

Note that mmap() is now used by default wherever possible, since this seems
to solve a number of problems. This default behavior can be changed in
config.h. Just to make sure this is clear, mmap() is not used for performance
reasons but to increase the number of systems on which dmidecode can be
successfully run. See the IA-64 subsection above for details.

CYGWIN

Dmidecode was reported to work under Cygwin. It seems that /dev/mem doesn't
work properly before version 1.5.10 though, so you will need to use at least
this version.


** MISCELLANEOUS TOOLS **

Three other tools come along with dmidecode: biosdecode, ownership and
vpddecode. These tools are only useful on systems with a BIOS, so they
are not built on IA-64 by default.

BIOSDECODE

This one prints all BIOS related information it can find in /dev/mem.
It used to be part of dmidecode itself, but as dmidecode was growing,
we felt that the non-DMI part had to be moved to a separate tool.

OWNERSHIP

This tool was written on a request by Luc Van de Velde for use with Novell
tools in his company. It retrieves the "ownership tag" that can be set on
most Compaq computers. Since it uses the same mechanisms dmidecode and
biosdecode use, and could be of some use for other people as well, we
decided to make it part of the project.

VPDDECODE

This tool prints the contents of the "vital product data" structure as
found in most IBM and Lenovo computers. It used to have a lookup table
for the machine name, but it was unreliable and hard to maintain so it
was ultimately dropped. It has a command line interface.


NEWS
Version 3.4 (Mon Jun 27 2022)
  - Support for SMBIOS 3.4.0. This includes new memory device types, new
    processor upgrades, new slot types and characteristics, decoding of memory
    module extended speed, new system slot types, new processor characteristics
    and new format of Processor ID.
  - Support for SMBIOS 3.5.0. This includes new processor upgrades, BIOS
    characteristics, new slot characteristics, new on-board device types, new
    pointing device interface types, and a new record type (type 45 -
    Firmware Inventory Information).
  - Decode HPE OEM records 194, 199, 203, 236, 237, 238 ans 240.
  - Bug fixes:
    Fix OEM vendor name matching
    Fix ASCII filtering of strings
    Fix crash with option -u
  - Minor improvements:
    Skip details of uninstalled memory modules
    Don't display the raw CPU ID in quiet mode
    Improve the formatting of the manual pages

Version 3.3 (Wed Oct 14 2020)
  - [BUILD] Allow overriding build settings from the environment.
  - [COMPATIBILITY] Document how the UUID fields are interpreted.
  - [PORTABILITY] Don't use memcpy on /dev/mem on arm64.
  - [PORTABILITY] Only scan /dev/mem for entry point on x86.
  - Support for SMBIOS 3.3.0. This includes new processor names, new port
    connector types, and new memory device form factors, types and
    technologies.
  - Add bios-revision, firmware-revision and system-sku-number to -s option.
  - Use the most appropriate unit for cache size.
  - Decode system slot base bus width and peers.
  - Important bug fixes:
    Fix Redfish Hostname print length
    Fix formatting of TPM table output
    Fix System Slot Information for PCIe SSD
    Don't choke on invalid processor voltage
  - Use the most appropriate unit for cache size.

Version 3.2 (Wed Sep 14 2018)
  - [COMPATIBILITY] The UUID is now displayed using lowercase letters, per
    RFC 4122 (#53569). You must ensure that any code parsing it is
    case-insensitive.
  - Support for SMBIOS 3.2.0. This includes new processor names, new socket
    and port connector types, new system slot state and property, and support
    for non-volatile memory (NVDIMM).
  - Support for Redfish management controllers.
  - A new command line option to query a specific structure by its handle.
  - A new command line option to query the system family string.
  - Support for 3 ThinkPad-specific structures (patch #9642).
  - Support for HPE's new company name.
  - Support UEFI on FreeBSD.
  - Important bug fixes:
    Fix firmware version of TPM device
    Fix the HPE UEFI feature flag check
  - (biosdecode) A new command line option to fully decode PIR information
    (support request #109339).

Version 3.1 (Tue May 23 2017)
  - Support for SMBIOS 3.1.0 and 3.1.1. This includes new chassis types, new
    processor family names, new processor family upgrade names, and new slot
    types, as well as support of larger BIOS ROM sizes and cache sizes, and a
    new structure type (43, TPM Device.)
  - A new command line option to query OEM strings.
  - All error messages are now printed on stderr (#47274, #48158.)
  - Several bug fixes related to 64-bit entry points (#50037 and more.)
  - Important bug fixes:
    #46176 (Unexpected end of file error)
    #46066 (Crash with SIGBUS)
  - Various minor fixes, improvements and cleanups.

Version 3.0 (Thu Sep 03 2015)
  - Support for SMBIOS 3.0. This includes new chassis types, new
    processor family names, new processor family upgrade names, new slot
    types, and new memory device types.
  - Support for the new 64-bit entry point (_SM3_) defined in SMBIOS 3.0.
  - Support for the new kernel interface (as of Linux v4.2) as an
    alternative to relying /dev/mem to access the entry point and DMI
    table.
  - Decoding of Acer-specific DMI type 170.
  - Decoding of HP-specific DMI types 212, 219 and 233.
  - Various minor fixes and output format cleanups.

Version 2.12 (Wed Apr 17 2013)
  - Support of the SMBIOS 2.8.0 specification.

Version 2.11 (Wed Jan 19 2011)
  - Support of the SMBIOS 2.7.0 specification:
    - UEFI support
    - Virtual machine flags in BIOS characteristics
    - Limited support for the Management Controller Host Interface
  - Various fixes that address stability.

Version 2.10 (Sun Nov 23 2008)
  - Support for Solaris (x86 only, of course).
  - Possibility to dump the SMBIOS/DMI table to a small binary file
    (option --dump-bin).
  - Possibility to read the SMBIOS/DMI table from such binary files
    (option --from-dump).
  - Support for SMBIOS 2.6. This includes new chassis types, new
    processor family names, new processor family upgrade names, bus
    address for system slots, and a new entry type for on-board devices,
    amongst many other minor changes.
  - Support for DMI entry type 31 (Boot integrity services).
  - Many processor family names taken from the CIM Schema document.
  - (vpddecode) No longer ask users to report broken records.
  - (vpddecode) Fix --quiet option.

Version 2.9 (Mon Feb 26 2007)
  - Support of the SMBIOS 2.5 specification. It adds many enumerated
    values for recent hardware, as well as CPU core and thread count
    reporting.
  - Decoding of 3 HP-specific entries. More vendor-specific entries can
    be supported later if vendors contribute code or documentation.
  - Run-time detection of EFI, so that a single binary can support
    Intel-based Macintosh machines and regular x86 machines.
  - Better IA-64 support.
  - Fixes to the decoding of individual fields, including the CPU
    signature of recent CPU models.
  - (biosdecode) Support of the FJKEYINF entry point type (for Fujitsu laptops).
  - (vpddecode) The product name look-up table was dropped. It was unreliable
    and a burden to maintain.
  - biosdecode, ownership and vpddecode are no longer built on IA-64.

Version 2.8 (Sat Feb 04 2006)
  - Option --string has four additional keywords available:
    system-uuid, chassis-type, processor-family and processor-frequency.

dmiopt.c
/*
 * Command line handling of dmidecode
 * This file is part of the dmidecode project.
 *
 *   Copyright (C) 2005-2008 Jean Delvare <jdelvare@suse.de>
 *
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation; either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program; if not, write to the Free Software
 *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
 */

#include <stdio.h>
#include <string.h>
#include <strings.h>
#include <stdlib.h>
#include <getopt.h>

#include "config.h"
#include "types.h"
#include "util.h"
#include "dmidecode.h"
#include "dmiopt.h"


/* Options are global */
struct opt opt;


/*
 * Handling of option --type
 */

struct type_keyword
{
	const char *keyword;
	const u8 *type;
};

static const u8 opt_type_bios[] = { 0, 13, 255 };
static const u8 opt_type_system[] = { 1, 12, 15, 23, 32, 255 };
static const u8 opt_type_baseboard[] = { 2, 10, 41, 255 };
static const u8 opt_type_chassis[] = { 3, 255 };
static const u8 opt_type_processor[] = { 4, 255 };
static const u8 opt_type_memory[] = { 5, 6, 16, 17, 255 };
static const u8 opt_type_cache[] = { 7, 255 };
static const u8 opt_type_connector[] = { 8, 255 };
static const u8 opt_type_slot[] = { 9, 255 };

static const struct type_keyword opt_type_keyword[] = {
	{ "bios", opt_type_bios },
	{ "system", opt_type_system },
	{ "baseboard", opt_type_baseboard },
	{ "chassis", opt_type_chassis },
	{ "processor", opt_type_processor },
	{ "memory", opt_type_memory },
	{ "cache", opt_type_cache },
	{ "connector", opt_type_connector },
	{ "slot", opt_type_slot },
};

static void print_opt_type_list(void)
{
	unsigned int i;

	fprintf(stderr, "Valid type keywords are:\n");
	for (i = 0; i < ARRAY_SIZE(opt_type_keyword); i++)
	{
		fprintf(stderr, "  %s\n", opt_type_keyword[i].keyword);
	}
}

static u8 *parse_opt_type(u8 *p, const char *arg)
{
	unsigned int i;

	/* Allocate memory on first call only */
	if (p == NULL)
	{
		p = (u8 *)calloc(256, sizeof(u8));
		if (p == NULL)
		{
			perror("calloc");
			return NULL;
		}
	}

	/* First try as a keyword */
	for (i = 0; i < ARRAY_SIZE(opt_type_keyword); i++)
	{
		if (!strcasecmp(arg, opt_type_keyword[i].keyword))
		{
			int j = 0;
			while (opt_type_keyword[i].type[j] != 255)
				p[opt_type_keyword[i].type[j++]] = 1;
			goto found;
		}
	}

	/* Else try as a number */
	while (*arg != '\0')
	{
		unsigned long val;
		char *next;

		val = strtoul(arg, &next, 0);
		if (next == arg || (*next != '\0' && *next != ',' && *next != ' '))
		{
			fprintf(stderr, "Invalid type keyword: %s\n", arg);
			print_opt_type_list();
			goto exit_free;
		}
		if (val > 0xff)
		{
			fprintf(stderr, "Invalid type number: %lu\n", val);
			goto exit_free;
		}

		p[val] = 1;
		arg = next;
		while (*arg == ',' || *arg == ' ')
			arg++;
	}

found:
	return p;

exit_free:
	free(p);
	return NULL;
}


/*
 * Handling of option --string
 */

/* This lookup table could admittedly be reworked for improved performance.
   Due to the low count of items in there at the moment, it did not seem
   worth the additional code complexity though. */
static const struct string_keyword opt_string_keyword[] = {
	{ "bios-vendor", 0, 0x04 },
	{ "bios-version", 0, 0x05 },
	{ "bios-release-date", 0, 0x08 },
	{ "bios-revision", 0, 0x15 },		/* 0x14 and 0x15 */
	{ "firmware-revision", 0, 0x17 },	/* 0x16 and 0x17 */
	{ "system-manufacturer", 1, 0x04 },
	{ "system-product-name", 1, 0x05 },
	{ "system-version", 1, 0x06 },
	{ "system-serial-number", 1, 0x07 },
	{ "system-uuid", 1, 0x08 },             /* dmi_system_uuid() */
	{ "system-sku-number", 1, 0x19 },
	{ "system-family", 1, 0x1a },
	{ "baseboard-manufacturer", 2, 0x04 },
	{ "baseboard-product-name", 2, 0x05 },
	{ "baseboard-version", 2, 0x06 },
	{ "baseboard-serial-number", 2, 0x07 },
	{ "baseboard-asset-tag", 2, 0x08 },
	{ "baseboard-location-in-chassis", 2, 0x0A },
	{ "chassis-manufacturer", 3, 0x04 },
	{ "chassis-type", 3, 0x05 },            /* dmi_chassis_type() */
	{ "chassis-version", 3, 0x06 },
	{ "chassis-serial-number", 3, 0x07 },
	{ "chassis-asset-tag", 3, 0x08 },
	{ "processor-family", 4, 0x06 },        /* dmi_processor_family() */
	{ "processor-manufacturer", 4, 0x07 },
	{ "processor-version", 4, 0x10 },
	{ "processor-frequency", 4, 0x16 },     /* dmi_processor_frequency() */
};

/* This is a template, 3rd field is set at runtime. */
static struct string_keyword opt_oem_string_keyword =
	{ NULL, 11, 0x00 };

static void print_opt_string_list(void)
{
	unsigned int i;

	fprintf(stderr, "Valid string keywords are:\n");
	for (i = 0; i < ARRAY_SIZE(opt_string_keyword); i++)
	{
		fprintf(stderr, "  %s\n", opt_string_keyword[i].keyword);
	}
}

static int parse_opt_string(const char *arg)
{
	unsigned int i;

	if (opt.string)
	{
		fprintf(stderr, "Only one string can be specified\n");
		return -1;
	}

	for (i = 0; i < ARRAY_SIZE(opt_string_keyword); i++)
	{
		if (!strcasecmp(arg, opt_string_keyword[i].keyword))
		{
			opt.string = &opt_string_keyword[i];
			return 0;
		}
	}

	fprintf(stderr, "Invalid string keyword: %s\n", arg);
	print_opt_string_list();
	return -1;
}

static int parse_opt_oem_string(const char *arg)
{
	unsigned long val;
	char *next;

	if (opt.string)
	{
		fprintf(stderr, "Only one string can be specified\n");
		return -1;
	}

	/* Return the number of OEM strings */
	if (strcmp(arg, "count") == 0)
		goto done;

	val = strtoul(arg, &next, 10);
	if (next == arg  || *next != '\0' || val == 0x00 || val > 0xff)
	{
		fprintf(stderr, "Invalid OEM string number: %s\n", arg);
		return -1;
	}

	opt_oem_string_keyword.offset = val;
done:
	opt.string = &opt_oem_string_keyword;
	return 0;
}

static u32 parse_opt_handle(const char *arg)
{
	u32 val;
	char *next;

	val = strtoul(arg, &next, 0);
	if (next == arg || *next != '\0' || val > 0xffff)
	{
		fprintf(stderr, "Invalid handle number: %s\n", arg);
		return ~0;
	}
	return val;
}

/*
 * Command line options handling
 */

/* Return -1 on error, 0 on success */
int parse_command_line(int argc, char * const argv[])
{
	int option;
	const char *optstring = "d:hqs:t:uH:V";
	struct option longopts[] = {
		{ "dev-mem", required_argument, NULL, 'd' },
		{ "help", no_argument, NULL, 'h' },
		{ "quiet", no_argument, NULL, 'q' },
		{ "string", required_argument, NULL, 's' },
		{ "type", required_argument, NULL, 't' },
		{ "dump", no_argument, NULL, 'u' },
		{ "dump-bin", required_argument, NULL, 'B' },
		{ "from-dump", required_argument, NULL, 'F' },
		{ "handle", required_argument, NULL, 'H' },
		{ "oem-string", required_argument, NULL, 'O' },
		{ "no-sysfs", no_argument, NULL, 'S' },
		{ "version", no_argument, NULL, 'V' },
		{ NULL, 0, NULL, 0 }
	};

	while ((option = getopt_long(argc, argv, optstring, longopts, NULL)) != -1)
		switch (option)
		{
			case 'B':
				opt.flags |= FLAG_DUMP_BIN;
				opt.dumpfile = optarg;
				break;
			case 'F':
				opt.flags |= FLAG_FROM_DUMP;
				opt.dumpfile = optarg;
				break;
			case 'd':
				opt.devmem = optarg;
				break;
			case 'h':
				opt.flags |= FLAG_HELP;
				break;
			case 'q':
				opt.flags |= FLAG_QUIET;
				break;
			case 's':
				if (parse_opt_string(optarg) < 0)
					return -1;
				opt.flags |= FLAG_QUIET;
				break;
			case 'O':
				if (parse_opt_oem_string(optarg) < 0)
					return -1;
				opt.flags |= FLAG_QUIET;
				break;
			case 't':
				opt.type = parse_opt_type(opt.type, optarg);
				if (opt.type == NULL)
					return -1;
				break;
			case 'H':
				opt.handle = parse_opt_handle(optarg);
				if (opt.handle  == ~0U)
					return -1;
				break;
			case 'u':
				opt.flags |= FLAG_DUMP;
				break;
			case 'S':
				opt.flags |= FLAG_NO_SYSFS;
				break;
			case 'V':
				opt.flags |= FLAG_VERSION;
				break;
			case '?':
				switch (optopt)
				{
					case 's':
						fprintf(stderr, "String keyword expected\n");
						print_opt_string_list();
						break;
					case 't':
						fprintf(stderr, "Type number or keyword expected\n");
						print_opt_type_list();
						break;
				}
				return -1;
		}

	/* Check for mutually exclusive output format options */
	if ((opt.string != NULL) + (opt.type != NULL)
	  + !!(opt.flags & FLAG_DUMP_BIN) + (opt.handle != ~0U) > 1)
	{
		fprintf(stderr, "Options --string, --type, --handle and --dump-bin are mutually exclusive\n");
		return -1;
	}

	if ((opt.flags & FLAG_FROM_DUMP) && (opt.flags & FLAG_DUMP_BIN))
	{
		fprintf(stderr, "Options --from-dump and --dump-bin are mutually exclusive\n");
		return -1;
	}

	return 0;
}

void print_help(void)
{
	static const char *help =
		"Usage: dmidecode [OPTIONS]\n"
		"Options are:\n"
		" -d, --dev-mem FILE     Read memory from device FILE (default: " DEFAULT_MEM_DEV ")\n"
		" -h, --help             Display this help text and exit\n"
		" -q, --quiet            Less verbose output\n"
		" -s, --string KEYWORD   Only display the value of the given DMI string\n"
		" -t, --type TYPE        Only display the entries of given type\n"
		" -H, --handle HANDLE    Only display the entry of given handle\n"
		" -u, --dump             Do not decode the entries\n"
		"     --dump-bin FILE    Dump the DMI data to a binary file\n"
		"     --from-dump FILE   Read the DMI data from a binary file\n"
		"     --no-sysfs         Do not attempt to read DMI data from sysfs files\n"
		"     --oem-string N     Only display the value of the given OEM string\n"
		" -V, --version          Display the version and exit\n";

	printf("%s", help);
}
